### PR TITLE
B 17642 resubmit sit request

### PR DESCRIPTION
**Pull request by TevinAdams**
AGILITY TICKET B-17642

**Summary**:
This pull request should allow a DOFSIT, DOASIT, and DDDSIT service item with model type "UpdateMTOServiceItemSIT" be submitted through the Patch UpdateMTOServiceItem endpoint in the PrimeAPI. This PR will also allow the following existing fields to be updated: Reason, SIT Departure Date, SIT Postal Code, and SIT Entry Date through the endpoint listed above. There were also 1 additional service item property added that is named "RequestApprovalsRequestedStatus". This value is a boolean that is initially set to false on MTOServiceItem creation. This value can be changed through the Patch request listed above.  Anytime a ServiceItem of model type UpdateMTOServiceItemSIT is passed through the endpoint listed above the status of the service item will be changed to "Submitted". In a future story validation will be added to reject the request if nothing was changed.

How to know if the service item you are testing with is a DOFSIT, DDDSIT or DOASIT item?
1. **Database**: In the database mto_service_items table the service item will have a re_service_id listed below.

Service Item Numbers:
DOFSIT - 998beda7-e390-4a83-b15e-578a24326937
DDDSIT - 5c80f3b5-548e-4077-9b8e-8d0390e73668
DOASIT - 05eb6ff1-5cf6-4918-b887-8260dda6b9fe

3. **Code**: Any service item with MTOServiceItem.ReServiceCode = "DOFSIT,DDDSIT or DOFSIT"

**MY TEST CASE:**
This endpoint was tested with swagger. 

*** Note if you want to  pass a param in if-match without commenting out the etag validation then here is how you find the correct if-match value. Put a console log in mto_service_item_updater.go for the encodedUpdatedAt variable on line 166. That value is the correct if-match value. But if you do not want to go through that just comment out the validation to test.

1. For simplicity the etag was by passed by commenting out its validation in file mto_service_item_updater.go lines 249-252. This was commented out because it would have taken much more time to understand the correct value to pass in here as i believe it is a generated hash that uses a third party library and a timestamp (Not completely sure). 
2. Put the  DOFSIT, DDDSIT or DOASIT service item ID in the mtoServiceItemID field
4. Make sure in the body of the request  the "modelType" attribute is set to "UpdateMTOServiceItemSIT".
5. Here is an example body that was used for testing.
{
  "modelType": "UpdateMTOServiceItemSIT",
  "sitRequestedDelivery": "2019-08-25",
  "timeMilitary2": "1400Z",
  "updateReason": "Test reason 3",
  "requestApprovalsRequestedStatus": true,
  "sitDepartureDate": "2019-08-25",
  "sitPostalCode" : "90210"
}

6. After executing the request check the database table mto_service_items and check the serviceitem you changed for the updated values.

### CODE REVIEW

**Comment by danieljordan-caci** [comment]
When loading up your code, the endpoint that this uses is the `PATCH` request to `/mto-service-items/{mtoServiceItemID}`, correct?

In your description you mention "Patch UpdateMTOServiceItemSIT endpoint in the PrimeAPI" but that isn't showing up in the swagger UI.

**Comment by danieljordan-caci** [comment]
When sending a PATCH request to `http://officelocal:3000/prime/v1/mto-service-items/e0638df8-5956-4420-aaab-b2187092fbce` with the body:
```
{
  "id": "e0638df8-5956-4420-aaab-b2187092fbce",
  "modelType": "UpdateMTOServiceItemSIT",
  "sitEntryDate": "2023-02-18"
}
```

I receive this:
```
{
  "detail": "Invalid input found while validating the service item.",
  "instance": "9807f118-c4e8-4abe-94fb-bb10114d93be",
  "title": "Validation Error",
  "invalidFields": {
    "status": [
      "cannot be updated"
    ]
  }
}
```

And when I try to run your exact test case json:
```
{
  "modelType": "UpdateMTOServiceItemSIT",
  "sitRequestedDelivery": "2019-08-25",
  "timeMilitary2": "1400Z",
  "updateReason": "Test reason 3",
  "requestApprovalsRequestedStatus": "true",
  "sitDepartureDate": "2019-08-25",
  "sitPostalCode" : "90210"
}
```

I get this error:
```
{
  "title": "Internal Server Error",
  "instance": "f0f35f6a-a1aa-4f1e-82c7-7711225612e9",
  "detail": "json: cannot unmarshal string into Go struct field .requestApprovalsRequestedStatus of type bool"
}
```

I haven't found a way to get a successful update to a SIT request yet. Let me know if I'm doing something incorrectly or hitting the wrong endpoint.

**Comment by TevinAdams** [comment]
> When loading up your code, the endpoint that this uses is the `PATCH` request to `/mto-service-items/{mtoServiceItemID}`, correct?
> 
> In your description you mention "Patch UpdateMTOServiceItemSIT endpoint in the PrimeAPI" but that isn't showing up in the swagger UI.



> When sending a PATCH request to `http://officelocal:3000/prime/v1/mto-service-items/e0638df8-5956-4420-aaab-b2187092fbce` with the body:
> 
> ```
> {
>   "id": "e0638df8-5956-4420-aaab-b2187092fbce",
>   "modelType": "UpdateMTOServiceItemSIT",
>   "sitEntryDate": "2023-02-18"
> }
> ```
> 
> I receive this:
> 
> ```
> {
>   "detail": "Invalid input found while validating the service item.",
>   "instance": "9807f118-c4e8-4abe-94fb-bb10114d93be",
>   "title": "Validation Error",
>   "invalidFields": {
>     "status": [
>       "cannot be updated"
>     ]
>   }
> }
> ```
> 
> And when I try to run your exact test case json:
> 
> ```
> {
>   "modelType": "UpdateMTOServiceItemSIT",
>   "sitRequestedDelivery": "2019-08-25",
>   "timeMilitary2": "1400Z",
>   "updateReason": "Test reason 3",
>   "requestApprovalsRequestedStatus": "true",
>   "sitDepartureDate": "2019-08-25",
>   "sitPostalCode" : "90210"
> }
> ```
> 
> I get this error:
> 
> ```
> {
>   "title": "Internal Server Error",
>   "instance": "f0f35f6a-a1aa-4f1e-82c7-7711225612e9",
>   "detail": "json: cannot unmarshal string into Go struct field .requestApprovalsRequestedStatus of type bool"
> }
> ```
> 
> I haven't found a way to get a successful update to a SIT request yet. Let me know if I'm doing something incorrectly or hitting the wrong endpoint.

Hey for my example the "True" needs to be True. I corrected it in the summary but thats supposed to be a boolean.  It failed because its a string.

**Comment by danieljordan-caci** [comment]
With that change, I am getting:
```
{
  "detail": "Invalid input found while validating the service item.",
  "instance": "52e5335c-ba73-4b58-968c-d95f66cbbc0a",
  "title": "Validation Error",
  "invalidFields": {
    "status": [
      "cannot be updated"
    ]
  }
}
```

**Comment by TevinAdams** [comment]
> With that change, I am getting:
> 
> ```
> {
>   "detail": "Invalid input found while validating the service item.",
>   "instance": "52e5335c-ba73-4b58-968c-d95f66cbbc0a",
>   "title": "Validation Error",
>   "invalidFields": {
>     "status": [
>       "cannot be updated"
>     ]
>   }
> }
> ```
> 
> Here is the entire PATCH:
> 
> ```
> curl -X 'PATCH' \
>   'http://officelocal:3000/prime/v1/mto-service-items/e0638df8-5956-4420-aaab-b2187092fbce' \
>   -H 'accept: application/json' \
>   -H 'If-Match: none' \
>   -H 'Content-Type: application/json' \
>   -H 'X-CSRF-Token: Jx1Xk7f7/JlpaGmoC6xJOQsfvSKMcCxWHcbpJPDQhM2Jj3YeD2+iinx+ePB9tayp1GTaFwS9pviVK1Yfg5p7BQ==' \
>   -d '{
> "modelType": "UpdateMTOServiceItemSIT",
> "sitRequestedDelivery": "2019-08-25",
> "timeMilitary2": "1400Z",
> "updateReason": "Test reason 3",
> "requestApprovalsRequestedStatus": true,
> "sitDepartureDate": "2019-08-25",
> "sitPostalCode" : "90210"
> }'
> ```

Your case and my case is passing for my test serviceItem. Is your serviceItem in REJECTED status? Im looking into why yours are working on mine but not yours but It must have something to do with the specific serviceItem.

**Comment by TevinAdams** [comment]
> When loading up your code, the endpoint that this uses is the `PATCH` request to `/mto-service-items/{mtoServiceItemID}`, correct?
> 
> In your description you mention "Patch UpdateMTOServiceItemSIT endpoint in the PrimeAPI" but that isn't showing up in the swagger UI.

Corrected the summary

**Comment by TevinAdams** [comment]
> Is there a reason only a DOFSIT can be the only service item that can be updated? I've copied the AC here:
> 
> ```
> Given a SIT PATCH request,
> When the SIT request is in a submitted or approved status 
> Then MilMove API blocks the request and sends back error saying status is not in an editable state
> 
> Given a SIT PATCH request,
> When the SIT request is in rejected status
> Then MilMove allows the request and updates status to SUBMITTED
> 
> Given a rejected SIT request
> When the GHC prime elects to re‐submit the existing request
> Then:
> The GHC Prime can update all parameters available on the initial SIT request. (We believe the Updates a service item by ID for a move by ID endpoint in the GHC API doc is used)
> The GHC Prime can request the SIT request be put in “approvals requested” status (believe the GHC prime will see it in a submitted status) --> we think is already existing functionality
> ```
> 
> These are the SIT-related service items: DOFSIT (Domestic origin 1st day SIT) DOASIT (Domestic origin add'l SIT) DOPSIT (Domestic origin SIT pickup) DDFSIT (Domestic destination1st day SIT) DDASIT (Domestic destination1st add'l SIT) DDPSIT (Domestic destination1st SIT pickup)
> 
> I'm not seeing anything in the AC that is talking about a DOFSIT in particular, so that's why I'm asking. The rest of the code looks good!

Thus far, the only model types that can be updated through this Patch request are:
DDDSIT - UpdateMTOServiceItemSIT
DOPSIT - UpdateMTOServiceItemSIT
DDSHUT - UpdateMTOServiceItemShuttle
DOSHUT - UpdateMTOServiceItemShuttle

image


DOPSIT is the initial and main service item that gets created and then the other ones (subServiceTypes)  like fuel surcharge and such are auto created off of the main SIT Item. The subServiceTypes are tied to the main one. When initially creating the story I believe it was always the impression of allowing edits to the main SIT Service Item of DOPSIT. I did bring that up at some point but as this is new to us all we may need to ask if they want us to open this endpoint to be able to update more serviceItems than just the ones shown above. I do not know. But I think by allowing more than the serviceItems shown above to be updated will be a big change. We will need to ask if that is wanted.

**Comment by TevinAdams** [comment]
> Is there a reason only a DOFSIT can be the only service item that can be updated? I've copied the AC here:
> 
> ```
> Given a SIT PATCH request,
> When the SIT request is in a submitted or approved status 
> Then MilMove API blocks the request and sends back error saying status is not in an editable state
> 
> Given a SIT PATCH request,
> When the SIT request is in rejected status
> Then MilMove allows the request and updates status to SUBMITTED
> 
> Given a rejected SIT request
> When the GHC prime elects to re‐submit the existing request
> Then:
> The GHC Prime can update all parameters available on the initial SIT request. (We believe the Updates a service item by ID for a move by ID endpoint in the GHC API doc is used)
> The GHC Prime can request the SIT request be put in “approvals requested” status (believe the GHC prime will see it in a submitted status) --> we think is already existing functionality
> ```
> 
> These are the SIT-related service items: DOFSIT (Domestic origin 1st day SIT) DOASIT (Domestic origin add'l SIT) DOPSIT (Domestic origin SIT pickup) DDFSIT (Domestic destination1st day SIT) DDASIT (Domestic destination1st add'l SIT) DDPSIT (Domestic destination1st SIT pickup)
> 
> I'm not seeing anything in the AC that is talking about a DOFSIT in particular, so that's why I'm asking. The rest of the code looks good!

Now allows this functionality for DOFSIT, DDDSIT and DOASIT service items.

**Comment by danieljordan-caci** [comment]
So in order to get this code to work, we have to comment out the if-match header in `mto_service_item_updater.go` on lines `248-251`. 

How would the prime know this? Is there a plan to figure this out in a future story? As of right now, with the if-match header staying in the request the AC is not met since the prime cannot comment out the code to bypass that check.

Other than that the code looks good and all of my tests are passing.

**Comment by TevinAdams** [comment]
> So in order to get this code to work, we have to comment out the if-match header in `mto_service_item_updater.go` on lines `248-251`.
> 
> How would the prime know this? Is there a plan to figure this out in a future story? As of right now, with the if-match header staying in the request the AC is not met since the prime cannot comment out the code to bypass that check.
> 
> Other than that the code looks good and all of my tests are passing.

The if-match is already an existing param and the code that is being added is not changing that. The prime doesn't have to change anything in terms of if-match.  Only thing the prime will need to do is add the additional params to the body that they wish to change. if-match is on other endpoints so the prime is already hitting these endpoints using the if-match. But again this code is not touching the if-match nor changing anything with it. Just for testing purposes it is easier to comment the if match check out when testing. We can spend time on trying to figure out what the if-match param is but that is outside of what this PR is doing. The prime doesn't have to do anything with the if-match.

**Comment by danieljordan-caci** [comment]
> > So in order to get this code to work, we have to comment out the if-match header in `mto_service_item_updater.go` on lines `248-251`.
> > How would the prime know this? Is there a plan to figure this out in a future story? As of right now, with the if-match header staying in the request the AC is not met since the prime cannot comment out the code to bypass that check.
> > Other than that the code looks good and all of my tests are passing.
> 
> The if-match is already an existing param and the code that is being added is not changing that. The prime doesn't have to change anything in terms of if-match. Only thing the prime will need to do is add the additional params to the body that they wish to change. if-match is on other endpoints so the prime is already hitting these endpoints using the if-match. But again this code is not touching the if-match nor changing anything with it. Just for testing purposes it is easier to comment the if match check out when testing. We can spend time on trying to figure out what the if-match param is but that is outside of what this PR is doing. The prime doesn't have to do anything with the if-match.

Based on the documentation here: https://transcom.github.io/mymove-docs/docs/adrs/optimistic-locking

The prime does have to include that value in the request. Looks like it's just a `base64` encoded version of the `updated_at` value. I think you're correct in saying that it isn't needed necessarily for this PR specifically, but the prime does have to include the `eTag` value in the `If-Match` value which ensures it is the most recent updated service item.

**Review by danieljordan-caci — APPROVED** [review]
LGTM! Thanks for all the clarification and work on this! Fun fun.

**Comment by TevinAdams** [comment]
> LGTM! Thanks for all the clarification and work on this! Fun fun.

No problem. This is an interesting system so happy for the conversations.

File: swagger-def/prime.yaml [hunk 1]
```diff
@@ -1684,6 +1684,25 @@ definitions:
...
+          updateReason:
+            type: string
+            description: Reason for updating service item.
+            x-nullable: true
+          sitPostalCode:
+            type: string
+            format: zip
+            example: '90210'
+            pattern: '^(\d{5}([\-]\d{4})?)$'
+            x-nullable: true
```
Comment (deandreJones) [inline swagger-def/prime.yaml:1696]: so the zip code is optional?

**Review by traskowskycaci — APPROVED** [review]
Was able to test locally and resubmit the sit request and confirm the db values matched my test values. Commented out the etag validation in order to successfully do so.

File: swagger-def/prime.yaml (hunk 1 above)
Comment (TevinAdams) [reply swagger-def/prime.yaml:1696]: @deandreJones Yes all the fields are optional. This is the patch request so zip code has been added as an optional field to update on the service item. The x-nullable: true does allow a null value but null will be treated as a no change and the zip code will not be changed.

**Review by cameroncaci — APPROVED** [review]

**Review by deandreJones — APPROVED** [review]

**Pull request merged**